code/data_structures/tree/binary_search_tree/binary_search_tree.cpp

```diff
@@ -1,88 +1,377 @@
 }
 
-//Postorder traversal
-void postorder(struct node *root)
+void BST::RemoveFrom(BinaryTreeNode *current, const int &value)
 {
-    if(root == NULL)
-        return;
-    postorder(root->left);
-    postorder(root->right);
-    cout<<root->key<<" ";
+	if (current != nullptr)
+	{
+		if (current->_value == value)
+			RemoveRoot();
+		else
+		{
+			if (value < current->_value && current->_Left != nullptr)
+			{
+				current->_Left->_value == value ?
+					RemoveMatch(current, current->_Left, true) :
+					RemoveFrom(current->_Left, value);
+			}
+			else if (value > current->_value && current->_Right != nullptr)
+			{
+				current->_Right->_value == value ?
+					RemoveMatch(current, current->_Right, false) :
+					RemoveFrom(current->_Left, value);
+			}
+			else
+				cout << "The value " << value << "wasn't found in the tree." << endl;
+		}
+	}
+	else
+		cout << "Tree is empty!" << endl;
 }
 
-//Insert key
-struct node* insert(struct node* node, int key)
+void BST::RemoveRoot(void)
 {
-    if (node == NULL) return newNode(key);
-    if (key < node->key)
-        node->left  = insert(node->left, key);
-    else if (key > node->key)
-        node->right = insert(node->right, key);
-    return node;
+	if (_Root != nullptr)
+	{
+		BinaryTreeNode *current = _Root;
+		int rootValue = _Root->_value;
+		int smallestInRightSubtree = 0;
+
+		// Case 0: The root has no children.
+		if (_Root->_Left == nullptr && _Root->_Right == nullptr)
+		{
+			_Root = nullptr;
+			delete current;
+			--_count;
+		}
+
+		// Case 1: The root has only one child.
+		else if (_Root->_Left != nullptr && _Root->_Right == nullptr)
+		{
+			_Root = _Root->_Left;
+			current->_Left = nullptr;
+			delete current;
+			--_count;
+			cout << "The root node with value " << rootValue << "was deleted." <<
+					"The new root contains value - " << _Root->_value << endl;
+		}
+		else if (_Root->_Left == nullptr && _Root->_Right != nullptr)
+		{
+			_Root = _Root->_Right;
+			current->_Right = nullptr;
+			delete current;
+			--_count;
+			cout << "The root node with value " << rootValue << "was deleted." <<
+					"The new root contains value - " << _Root->_value << endl;
+		}
+
+		//Case 2: The root has two children.
+		else 
+		{
+			smallestInRightSubtree = FindSmallestFrom(_Root->_Right);
+			RemoveFrom(_Root, smallestInRightSubtree);
+			_Root->_value = smallestInRightSubtree;
+			--_count;
+			cout << "The root containing value " << rootValue <<
+					" was overwritten with " << _Root->_value << endl;
+		}
+	}
+	else
+	{
+		cout << "The tree is empty!" << endl;
+	}
+}
+
+void BST::RemoveMatch(BinaryTreeNode *parent, BinaryTreeNode *match, bool left)
+{
+	if (_Root != nullptr)
+	{
+		BinaryTreeNode *current = new BinaryTreeNode();
+		int matchValue = match->_value;
+		int smallestInRightSubtree = 0;
+
+		//Case 0 : The match node have no children.
+		if (match->_Left == nullptr && match->_Right == nullptr)
+		{
+			current = match;
+			left == true ? parent->_Left == nullptr : 
```

Comment:
You’re checking if true == true? Why?

Reply:
This bool indicates how the matching node relates to its parent node (left/right). 
Look at RemoveFrom function.